[masonry] Handle masonry check if it is a grid box

stwrt opened the pull request:
#### 2905a83856e471c1d1877a715c21f911afcacf8c
<pre>
[masonry] Handle masonry check if it is a grid box
<a href="https://bugs.webkit.org/show_bug.cgi?id=302263">https://bugs.webkit.org/show_bug.cgi?id=302263</a>
<a href="https://rdar.apple.com/problem/164404373">rdar://problem/164404373</a>

Reviewed by Tim Nguyen and Elika Etemad.

We were not properly checking masonry display type in the RenderStyleInlines.
This fixes 37 test cases.

* LayoutTests/TestExpectations
* Source/WebCore/rendering/style/RenderStyleInlines.h

Canonical link: <a href="https://commits.webkit.org/302819@main">https://commits.webkit.org/302819@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/11eef4b1bb92ed85c38a12d63a0a3eb18c05da19

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/130256 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/2527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/41210 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/137674 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/81822 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/0eb6904d-8598-45ed-be17-0515e2bd43a9) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/132127 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/2527 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/2418 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/99239 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/67083 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a9307612-edf9-4d9d-8312-cad61c4402e0) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/133203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/1864 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/116652 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/79932 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/c34198ef-5594-4891-a37b-90896480c0fb) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/1785 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/34780 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/80930 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/110329 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/35287 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/140151 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/2317 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/2149 "Found 2 new test failures: imported/w3c/web-platform-tests/webvtt/rendering/cues-with-video/processing-model/selectors/cue/white-space_pre.html media/media-source/media-source-interruption-with-resume-allowing-play.html (failure)") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/107754 "Failure limit exceed. At least found 1 new test failure: imported/w3c/web-platform-tests/event-timing/interactionid-orphan-pointerup.html (failure)") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/2361 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/112995 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/107653 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27416 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/1826 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/31445 "Passed tests") | [❌ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/55270 "Hash 11eef4b1 for PR 53678 does not build (failure)") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/2387 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/65774 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/2204 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/2408 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/2313 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

stwrt commented on Source/WebCore/rendering/style/RenderStyleInlines.h:
```diff
@@ -635,7 +635,7 @@ constexpr bool RenderStyle::isDisplayDeprecatedFlexibleBox() const { return isDi
 constexpr bool RenderStyle::isDisplayFlexibleOrGridBox() const { return isDisplayFlexibleOrGridBox(display()); }
 constexpr bool RenderStyle::isDisplayFlexibleOrGridBox(DisplayType display) { return isDisplayFlexibleBox(display) || isDisplayGridBox(display); }
-constexpr bool RenderStyle::isDisplayGridBox(DisplayType display) { return display == DisplayType::Grid || display == DisplayType::InlineGrid; }
+constexpr bool RenderStyle::isDisplayGridBox(DisplayType display) { return display == DisplayType::Grid || display == DisplayType::InlineGrid || display == DisplayType::Masonry || display == DisplayType::InlineMasonry; }
```
I think we may want to have a separate `isDisplayMasonryBox` and then just put it in the same places we are checking for grid.

Would appreciate feedback on this.

stwrt replied:
Looking at other spots in the code base I see this similar pattern is used multiple times.

nt1m replied:
I think this makes sense, given masonry is going to be renamed to something-grid

sammygill replied:
Are we sure that's what we want? To me it seems like these functions are trying to check if the display type of these boxes establish a grid container like what is described in: https://www.w3.org/TR/css-display-3/#propdef-display

I am not sure what the status is regarding what effect the new display type value is supposed to have but from what I can see https://drafts.csswg.org/css-grid-3/#masonry-switch says that it is supposed to establish a masonry container.

Not sure the relationship between the two but from my reading it seems like they are two different things. Maybe that's ok and I'm just missing some info

alanbaradlay replied:
I think at this granularity it's more about how our code is structured than what the spec says. If RenderGrid represents a masonry container and masonry layout is simply a variant handled internally by the grid renderer, then it's perfectly fine to leave it as is.

fantasai replied:
I agree with Alan. If we need to distinguish somewhere, we could have separate `isDisplayGridGrid()` vs `isDisplayGridMasonry()` to be more specific. Or whatever. But these are both generating a grid formatting context of some kind, and they're current implemented using the same RenderObject subtype, so this seems fine to me for now.

stwrt committed:
[masonry] Handle masonry check if it is a grid box
https://bugs.webkit.org/show_bug.cgi?id=302263
rdar://problem/164404373

Reviewed by Tim Nguyen and Elika Etemad.

We were not properly checking masonry display type in the RenderStyleInlines.
This fixes 37 test cases.

* LayoutTests/TestExpectations
* Source/WebCore/rendering/style/RenderStyleInlines.h

Canonical link: https://commits.webkit.org/302819@main
